From 5c3b29a43cfdf32269cefb83e63299ef0b9a49cb Mon Sep 17 00:00:00 2001
From: "neil.shi" <neil.shi@quectel.com>
Date: Thu, 1 Jun 2023 10:22:05 +0800
Subject: [PATCH 044/188] wifi: ath11k: add support for unit test command

For debug purpose, FW need to execute some unit test command.
This patch support this requirement and user can file unit test
command via iw.

Signed-off-by: neil.shi <neil.shi@quectel.com>
[eugen.hristev@collabora.com: port to 6.8]
Cherry-picked-from: 7b7f47d7c356 ("Subject: [PATCH 05/42] wifi: ath11k: add support for unit test command")
Signed-off-by: Eugen Hristev <eugen.hristev@collabora.com>
Signed-off-by: Cristian Ciocaltea <cristian.ciocaltea@collabora.com>
---
 drivers/net/wireless/ath/ath11k/Makefile  |  3 +-
 drivers/net/wireless/ath/ath11k/mac.c     |  7 ++
 drivers/net/wireless/ath/ath11k/unitest.c | 96 +++++++++++++++++++++++
 drivers/net/wireless/ath/ath11k/unitest.h | 44 +++++++++++
 drivers/net/wireless/ath/ath11k/wmi.c     | 43 ++++++++++
 drivers/net/wireless/ath/ath11k/wmi.h     | 25 ++++++
 6 files changed, 217 insertions(+), 1 deletion(-)
 create mode 100644 drivers/net/wireless/ath/ath11k/unitest.c
 create mode 100644 drivers/net/wireless/ath/ath11k/unitest.h

diff --git a/drivers/net/wireless/ath/ath11k/Makefile b/drivers/net/wireless/ath/ath11k/Makefile
index b1435fcf3..ecf04d400 100644
--- a/drivers/net/wireless/ath/ath11k/Makefile
+++ b/drivers/net/wireless/ath/ath11k/Makefile
@@ -19,7 +19,8 @@ ath11k-y += core.o \
 	    hw.o \
 	    pcic.o \
 	    fw.o \
-	    p2p.o
+	    p2p.o\
+	    unitest.o
 
 ath11k-$(CONFIG_ATH11K_DEBUGFS) += debugfs.o debugfs_htt_stats.o debugfs_sta.o
 ath11k-$(CONFIG_NL80211_TESTMODE) += testmode.o
diff --git a/drivers/net/wireless/ath/ath11k/mac.c b/drivers/net/wireless/ath/ath11k/mac.c
index 2d55cdc4d..6133d206a 100644
--- a/drivers/net/wireless/ath/ath11k/mac.c
+++ b/drivers/net/wireless/ath/ath11k/mac.c
@@ -24,6 +24,7 @@
 #include "debugfs_sta.h"
 #include "hif.h"
 #include "wow.h"
+#include "unitest.h"
 
 #define CHAN2G(_channel, _freq, _flags) { \
 	.band                   = NL80211_BAND_2GHZ, \
@@ -255,6 +256,10 @@ static const u32 ath11k_smps_map[] = {
 	[WLAN_HT_CAP_SM_PS_DISABLED] = WMI_PEER_SMPS_PS_NONE,
 };
 
+static struct wiphy_vendor_command ath11k_vendor_cmds[] = {
+	ath11k_unit_test_command,
+};
+
 enum nl80211_he_ru_alloc ath11k_mac_phy_he_ru_to_nl80211_he_ru_alloc(u16 ru_phy)
 {
 	enum nl80211_he_ru_alloc ret;
@@ -10606,6 +10611,8 @@ static int __ath11k_mac_register(struct ath11k *ar)
 	ar->hw->wiphy->iftype_ext_capab = ath11k_iftypes_ext_capa;
 	ar->hw->wiphy->num_iftype_ext_capab =
 		ARRAY_SIZE(ath11k_iftypes_ext_capa);
+	ar->hw->wiphy->vendor_commands = ath11k_vendor_cmds;
+	ar->hw->wiphy->n_vendor_commands = ARRAY_SIZE(ath11k_vendor_cmds);
 
 	if (ar->supports_6ghz) {
 		wiphy_ext_feature_set(ar->hw->wiphy,
diff --git a/drivers/net/wireless/ath/ath11k/unitest.c b/drivers/net/wireless/ath/ath11k/unitest.c
new file mode 100644
index 000000000..541925af0
--- /dev/null
+++ b/drivers/net/wireless/ath/ath11k/unitest.c
@@ -0,0 +1,96 @@
+// SPDX-License-Identifier: BSD-3-Clause-Clear
+/**
+ * Copyright (c) 2020 The Linux Foundation. All rights reserved.
+ */
+
+#include <net/netlink.h>
+#include "debug.h"
+#include "unitest.h"
+
+const struct nla_policy ath11k_unitest_policy[UNITEST_MAX + 1] = {
+	[UNITEST_MODULE_ID] = {.type = NLA_U32},
+	[UNITEST_ARGS_NUM] = {.type = NLA_U32},
+	[UNITEST_ARGS] = {.type = NLA_BINARY,
+		.len = MAX_UNITEST_MEMORY_LEN},
+};
+
+/**
+ * ath11k_unit_test() - send unit test command
+ * @wiphy:    wiphy structure pointer
+ * @wdev:     Wireless device structure pointer
+ * @data:     Pointer to the data received
+ * @data_len: Length of @data
+ *
+ * Return: 0 on success; errno on failure
+ */
+
+int ath11k_unit_test(struct wiphy *wiphy, struct wireless_dev *wdev,
+		     const void *data, int data_len)
+{
+	struct ieee80211_hw *hw = wiphy_to_ieee80211_hw(wiphy);
+	struct ath11k *ar = hw->priv;
+	struct ath11k_base *ab = ar->ab;
+	struct nlattr *tb[UNITEST_MAX + 1];
+	int status = 0;
+	u32 module_id, num_args, temp_num_args, len;
+	bool sta_found = false;
+	struct ath11k_vif *arvif;
+	struct unit_test_cmd utest_cmd = {0};
+
+	list_for_each_entry(arvif, &ar->arvifs, list) {
+		if (arvif->vdev_type != WMI_VDEV_TYPE_STA ||
+		    arvif->vdev_subtype != WMI_VDEV_SUBTYPE_NONE)
+			continue;
+		sta_found = true;
+		break;
+	}
+	if (!sta_found) {
+		ath11k_warn(ar->ab, "no sta found.");
+		return -EINVAL;
+	}
+
+	utest_cmd.vdev_id = arvif->vdev_id;
+
+	if (nla_parse(tb, UNITEST_MAX, data, data_len, ath11k_unitest_policy, NULL)) {
+		ath11k_warn(ab, "Invalid ATTR");
+		return -EINVAL;
+	}
+
+	if (!tb[UNITEST_MODULE_ID]) {
+		ath11k_warn(ab, "attr unitest module id failed");
+		return -EINVAL;
+	}
+	module_id = nla_get_u32(tb[UNITEST_MODULE_ID]);
+	utest_cmd.module_id = module_id;
+
+	if (!tb[UNITEST_ARGS_NUM]) {
+		ath11k_warn(ab, "attr unitest args num failed");
+		return -EINVAL;
+	}
+	num_args = nla_get_u32(tb[UNITEST_ARGS_NUM]);
+	utest_cmd.num_args = num_args;
+	if (num_args > UNIT_TEST_MAX_NUM_ARGS) {
+		ath11k_warn(ab, "num args exceed");
+		return -EINVAL;
+	}
+
+	if (!tb[UNITEST_ARGS]) {
+		ath11k_warn(ab, "attr unitest args failed");
+		return -EINVAL;
+	}
+	len = nla_len(tb[UNITEST_ARGS]);
+	temp_num_args = len / sizeof(u32);
+	if (num_args != temp_num_args) {
+		ath11k_warn(ab, "num args mismatch");
+		return -EINVAL;
+	}
+	nla_memcpy(utest_cmd.args, tb[UNITEST_ARGS], len);
+
+	status = ath11k_wmi_set_unit_test(ar, &utest_cmd);
+	if (status) {
+		ath11k_warn(ab, "Unable to post unit test message (status-%d)", status);
+		return -EINVAL;
+	}
+
+	return 0;
+}
diff --git a/drivers/net/wireless/ath/ath11k/unitest.h b/drivers/net/wireless/ath/ath11k/unitest.h
new file mode 100644
index 000000000..9fee1cd68
--- /dev/null
+++ b/drivers/net/wireless/ath/ath11k/unitest.h
@@ -0,0 +1,44 @@
+/* SPDX-License-Identifier: BSD-3-Clause-Clear */
+/**
+ * Copyright (c) 2020  The Linux Foundation. All rights reserved.
+ */
+
+#ifndef _UNIT_TEST_H_
+#define _UNIT_TEST_H_
+
+#define VENDOR_ID	0x001374
+#define QCA_NL80211_VENDOR_UNITEST_SUBCMD	84
+#define MAX_UNITEST_MEMORY_LEN	128
+
+enum qca_wlan_vendor_attr_unit_test {
+	QCA_WLAN_VENDOR_ATTR_UNIT_TEST_INVALID = 0,
+	QCA_WLAN_VENDOR_ATTR_UNIT_TEST_MODULE_ID,
+	QCA_WLAN_VENDOR_ATTR_UNIT_TEST_ARGS_NUM,
+	QCA_WLAN_VENDOR_ATTR_UNIT_TEST_ARGS,
+	/* keep last */
+	QCA_WLAN_VENDOR_ATTR_UNIT_TEST_AFTER_LAST,
+	QCA_WLAN_VENDOR_ATTR_UNIT_TEST_MAX =
+		QCA_WLAN_VENDOR_ATTR_UNIT_TEST_AFTER_LAST - 1,
+};
+
+#define UNITEST_INVALID	QCA_WLAN_VENDOR_ATTR_UNIT_TEST_INVALID
+#define UNITEST_MODULE_ID	QCA_WLAN_VENDOR_ATTR_UNIT_TEST_MODULE_ID
+#define UNITEST_ARGS	QCA_WLAN_VENDOR_ATTR_UNIT_TEST_ARGS
+#define UNITEST_ARGS_NUM	QCA_WLAN_VENDOR_ATTR_UNIT_TEST_ARGS_NUM
+#define UNITEST_MAX	QCA_WLAN_VENDOR_ATTR_UNIT_TEST_MAX
+
+#define ath11k_unit_test_command \
+{\
+	.info.vendor_id = VENDOR_ID,\
+	.info.subcmd = QCA_NL80211_VENDOR_UNITEST_SUBCMD,\
+	.flags = WIPHY_VENDOR_CMD_NEED_WDEV |\
+		WIPHY_VENDOR_CMD_NEED_NETDEV |\
+		WIPHY_VENDOR_CMD_NEED_RUNNING,\
+	.doit = ath11k_unit_test,\
+	.policy = VENDOR_CMD_RAW_DATA,\
+	.maxattr = QCA_WLAN_VENDOR_ATTR_UNIT_TEST_MAX,\
+}
+
+int ath11k_unit_test(struct wiphy *wiphy, struct wireless_dev *wdev,
+		     const void *data, int data_len);
+#endif
diff --git a/drivers/net/wireless/ath/ath11k/wmi.c b/drivers/net/wireless/ath/ath11k/wmi.c
index 2ca065da1..98d3ccf92 100644
--- a/drivers/net/wireless/ath/ath11k/wmi.c
+++ b/drivers/net/wireless/ath/ath11k/wmi.c
@@ -10142,3 +10142,46 @@ bool ath11k_wmi_supports_6ghz_cc_ext(struct ath11k *ar)
 	return test_bit(WMI_TLV_SERVICE_REG_CC_EXT_EVENT_SUPPORT,
 			ar->ab->wmi_ab.svc_map) && ar->supports_6ghz;
 }
+
+int ath11k_wmi_set_unit_test(struct ath11k *ar, struct unit_test_cmd *unit_test)
+{
+	struct ath11k_pdev_wmi *wmi = ar->wmi;
+	struct sk_buff *skb;
+	struct wmi_unit_test_cmd_fixed_param *cmd;
+	u32 len, args_tlv_len;
+	u8 *buf_ptr;
+	u32 *args;
+	struct wmi_tlv *tlv;
+	u32 i;
+
+	args_tlv_len = TLV_HDR_SIZE + unit_test->num_args * sizeof(u32);
+
+	len = sizeof(struct wmi_unit_test_cmd_fixed_param) + args_tlv_len;
+	skb = ath11k_wmi_alloc_skb(wmi->wmi_ab, len);
+	if (!skb)
+		return -ENOMEM;
+
+	cmd = (struct wmi_unit_test_cmd_fixed_param *)skb->data;
+	buf_ptr = (u8 *)cmd;
+	cmd->tlv_header = FIELD_PREP(WMI_TLV_TAG,
+				     WMI_TAG_UNIT_TEST_CMD) |
+				     FIELD_PREP(WMI_TLV_LEN, sizeof(*cmd) - TLV_HDR_SIZE);
+	cmd->vdev_id = unit_test->vdev_id;
+	cmd->module_id = unit_test->module_id;
+	cmd->num_args = unit_test->num_args;
+
+	buf_ptr += sizeof(*cmd);
+
+	tlv  = (struct wmi_tlv *)buf_ptr;
+	tlv->header = FIELD_PREP(WMI_TLV_TAG, WMI_TAG_ARRAY_UINT32) |
+				 FIELD_PREP(WMI_TLV_LEN, unit_test->num_args * sizeof(u32));
+	args = (u32 *)(buf_ptr + TLV_HDR_SIZE);
+	ath11k_info(ar->ab, "module id = 0x%x, num args = %u",
+		    unit_test->module_id, unit_test->num_args);
+	for (i = 0; (i < unit_test->num_args && i < UNIT_TEST_MAX_NUM_ARGS); i++) {
+		args[i] = unit_test->args[i];
+		ath11k_info(ar->ab, "0x%x,", unit_test->args[i]);
+	}
+
+	return ath11k_wmi_cmd_send(wmi, skb, WMI_UNIT_TEST_CMDID);
+}
diff --git a/drivers/net/wireless/ath/ath11k/wmi.h b/drivers/net/wireless/ath/ath11k/wmi.h
index 51f57423a..759f0b92d 100644
--- a/drivers/net/wireless/ath/ath11k/wmi.h
+++ b/drivers/net/wireless/ath/ath11k/wmi.h
@@ -6438,6 +6438,29 @@ enum wmi_wmm_params_type {
 
 const void **ath11k_wmi_tlv_parse_alloc(struct ath11k_base *ab,
 					struct sk_buff *skb, gfp_t gfp);
+
+#define UNIT_TEST_MAX_NUM_ARGS    8
+
+struct unit_test_cmd {
+	u32 vdev_id;
+	u32 module_id;
+	u32 num_args;
+	u32 args[UNIT_TEST_MAX_NUM_ARGS];
+};
+
+struct wmi_unit_test_cmd_fixed_param {
+	u32 tlv_header;
+	u32 vdev_id;
+	u32 module_id;
+	u32 num_args;
+	u32 diag_token;
+	/**
+	 * TLV (tag length value) parameters follow the wmi_unit_test_cmd_fixed_param
+	 * structure. The TLV's are:
+	 *     u32 args[];
+	 */
+} __packed;
+
 int ath11k_wmi_cmd_send(struct ath11k_pdev_wmi *wmi, struct sk_buff *skb,
 			u32 cmd_id);
 struct sk_buff *ath11k_wmi_alloc_skb(struct ath11k_wmi_base *wmi_sc, u32 len);
@@ -6630,4 +6653,6 @@ int ath11k_wmi_send_vdev_set_tpc_power(struct ath11k *ar,
 int ath11k_wmi_peer_set_cfr_capture_conf(struct ath11k *ar,
 					 u32 vdev_id, const u8 *mac,
 					 struct wmi_peer_cfr_capture_conf_arg *arg);
+int ath11k_wmi_set_unit_test(struct ath11k *ar, struct unit_test_cmd *unit_test);
+
 #endif
-- 
2.55.0

